imagecraft/services/pack.py

```diff
@@ -43,15 +49,63 @@ def __init__(
         self._build_plan = build_plan
 
     @override
-    def pack(self, prime_dir: pathlib.Path, dest: pathlib.Path) -> list[pathlib.Path]:  # noqa: ARG002
+    def pack(self, prime_dir: Path, dest: Path) -> list[Path]:  # noqa: ARG002
         """Pack the image.
 
         :param prime_dir: Directory path to the prime directory.
         :param dest: Directory into which to write the package(s).
         :returns: A list of paths to created packages.
         """
+        # Pydantic has already validated that there is only a single volume before now
+        project = cast(Project, self._project)
+        if len(project.volumes) != 1:
+            raise AssertionError("This code can only handle one volume")
+        volume_name, volume = next(iter(project.volumes.items()))
+        disk_image_file = dest / (volume_name + os.extsep + "img")
+
+        # Create empty image
+        gptutil.create_empty_gpt_image(
+            imagepath=disk_image_file,
+            sector_size=SECTOR_SIZE,
+            layout=volume,
+        )
+
+        # Create partition images with filesystems.  These are always recreated, but we
+        # may want to revisit that once this is solved:
+        # https://github.com/canonical/craft-parts/issues/665
+        project_dirs = self._services.lifecycle.project_info.dirs
+        with tempfile.TemporaryDirectory() as partition_dir:
+            for structure_item in volume.structure:
+                partition_name = get_partition_name(volume_name, structure_item)
+                partition_prime_dir = project_dirs.get_prime_dir(
+                    partition=partition_name
+                )
+                partition_img = (
+                    Path(partition_dir) / f"{volume_name}.{structure_item.name}.img"
```

Comment:
Calling this `partition_dir` could be confusing, my first reaction was to think it was the top-level partitions work dir. I would rather leave it as `temp_dir` or something like that.